stdlib/public/runtime/Paths.cpp

```diff
@@ -0,0 +1,196 @@
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2022 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// Functions that obtain paths that might be useful within the runtime.
+//
+//===----------------------------------------------------------------------===//
+
+#include "swift/Runtime/Debug.h"
+#include "swift/Runtime/Paths.h"
+#include "swift/Threading/Once.h"
+
+#if !defined(_WIN32) || defined(__CYGWIN__)
+#include <dlfcn.h>
+#else
+#define WIN32_LEAN_AND_MEAN
+#define NOMINMAX
+#include <windows.h>
+#endif
+
+#include <cstdlib>
+#include <cstring>
+
+namespace {
+
+swift::once_t runtimePathToken;
+const char *runtimePath;
+
+swift::once_t rootPathToken;
+const char *rootPath;
+
+void _swift_initRuntimePath(void *);
+void _swift_initRootPath(void *);
+
+inline bool _swift_isPathSep(char ch) {
+#ifdef _WIN32
+  return ch == '/' || ch == '\\';
+#else
+  return ch == '/';
+#endif
+}
+
+#if !defined(_WIN32) || defined(__CYGWIN__)
+#define PATHSEP_STR "/"
+#define PATHSEP_CHR '/'
+#else
+#define PATHSEP_STR "\\"
+#define PATHSEP_CHR '\\'
+#endif
+
+#ifdef _WIN32
+#define EXE_EXT ".exe"
+#else
+#define EXE_EXT ""
+#endif
+
+}
+
+SWIFT_RUNTIME_EXPORT
+const char *
+swift_getRuntimePath()
+{
+  swift::once(runtimePathToken, _swift_initRuntimePath, nullptr);
+  return runtimePath;
+}
+
+SWIFT_RUNTIME_EXPORT
+const char *
+swift_getRootPath()
+{
+  swift::once(rootPathToken, _swift_initRootPath, nullptr);
+  return rootPath;
+}
+
+namespace {
+
+void
+_swift_initRootPath(void *)
+{
+  // SWIFT_ROOT overrides the path returned by this function
+  char *swiftRoot = getenv("SWIFT_ROOT");
+  if (swiftRoot) {
+    size_t len = std::strlen(swiftRoot);
+    if (len) {
+      // Make sure it has a trailing path separator
+      if (_swift_isPathSep(swiftRoot[len - 1]))
+        rootPath = ::strdup(swiftRoot);
+      else {
+        char *path = (char *)std::malloc(len + 2);
+        std::memcpy(path, swiftRoot, len);
+        std::memcpy(path + len, PATHSEP_STR, 2);
+        rootPath = path;
+      }
+      return;
+    }
+  }
+
+  // Otherwise, get the runtime path and remove /lib/libswiftCore.(dylib|dll|so)
+  const char *runtimePath = swift_getRuntimePath();
+  const char *end = runtimePath + std::strlen(runtimePath);
+  const char *ptr = end;
+  int count = 0;
+
+  while (ptr > runtimePath) {
+    int ch = *--ptr;
+    if (_swift_isPathSep(ch)) {
+      if (++count == 2)
+        break;
+    }
+  }
+
+  if (count != 2) {
+    swift::fatalError(/* flags = */ 0,
+                      "libswiftCore should be installed in a `lib` "
+                      "directory: %s\n",
+                      runtimePath);
+  }
```

Comment:
(This is exactly why I asked you to review it :-))

So, on Windows, if we want to get at another file from the runtime, where would it be relative to the runtime? The point of having these functions is to allow us to do that in a system-independent manner, so if Windows needs to do something slightly different, that's just fine.

Reply:
I think that is the problem, the default installation does use a unix-ish layout, but that isn't guaranteed.  It should support the flat tree (application co-installed redistribution) and the system installed layout.

Reply:
Seems like we should tolerate anything, and fail gracefully if we can't understand the layout we're in.

Reply:
I've made it more flexible. If it finds `libswiftCore` in a `lib` directory, it will remove the `lib` part as well, otherwise it just returns the directory it found `libswiftCore` in.